sql/parser/builtins.go

```diff
@@ -943,6 +961,108 @@ func datumToString(datum Datum) (string, error) {
 	return "", fmt.Errorf("argument type unsupported: %s", datum.Type())
 }
 
+var replaceSubRe = regexp.MustCompile(`\\[&1-9]`)
+
+func regexpReplace(s, pattern, to, sqlFlags string) (Datum, error) {
+	pattern, global, err := regexpEvalFlags(pattern, sqlFlags)
+	if err != nil {
+		return nil, err
+	}
+
+	patternRe, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	matchCount := 1
+	if global {
+		matchCount = -1
+	}
+
+	replaceIndex := 0
+	var newString bytes.Buffer
+
+	// regexp.ReplaceAllStringFunc cannot be used here because it does not provide
+	// access to regexp submatches for expansion in the replacement string.
+	// regexp.ReplaceAllString cannot be used here because it does not allow
+	// replacement of a specific number of matches, and does not expose the full
+	// match for expansion in the replacement string.
+	for _, matchIndex := range patternRe.FindAllStringSubmatchIndex(s, matchCount) {
+		start := matchIndex[0]
+		end := matchIndex[1]
+
+		preMatch := s[replaceIndex:start]
+		newString.WriteString(preMatch)
+
+		match := s[start:end]
+		matchTo := replaceSubRe.ReplaceAllStringFunc(to, func(repl string) string {
+			subRef := repl[len(repl)-1]
+			if subRef == '&' {
+				return match
+			}
+
+			sub, _ := strconv.Atoi(string(subRef))
```

Comment:
The `string` cast is needed because `subRef` is a `byte` due to the string indexing. There should never be an error because `repl`'s last char will always be either `&` of `1-9` (see `replaceSubRe`). Should I assert this to be true by panicking in the case of an error?